proposals/tf-operator-design-v1alpha2.md

```diff
@@ -0,0 +1,435 @@
+type TFControllerConfiguration struct {
+    // ReconcilerSyncLoopPeriod is the amount of time the reconciler sync states loop
+    // wait between two reconciler sync.
+    // It is set to 15 sec by default.
+    // TODO(cph): maybe we can let it grows by multiple in the future
+    // and up to 5 minutes to reduce idle loop.
+    // e.g. 15s, 30s, 60s, 120s...
+    ReconcilerSyncLoopPeriod metav1.Duration
+}
+
+// DefaultTFControllerConfiguration is the suggested tf-operator configuration for production.
+var DefaultTFControllerConfiguration TFControllerConfiguration = TFControllerConfiguration{
+    ReconcilerSyncLoopPeriod: 15 * time.Second,
+}
+```
+
+Reconciler use a `ReconcilerSyncLoopPeriod` to determine whether we should call this reconciler or ignore it. We should leave a record `LastReconcileTime` in TFJob object of course:
+
+```go
+    // Represents last time when the TFJob was reconciled. It is not guaranteed to
+    // be set in happens-before order across separate operations.
+    // It is represented in RFC3339 form and is in UTC.
+    LastReconcileTime *metav1.Time `json:"lastReconcileTime,omitempty"`
+```
+
+As `tfJobImformer` provides a forcing resync mechanism by calling `UpdateFunc` which defined in `ResourceEventHandlerFuncs` periodically. We can call the reconciler in this function:
+
+- UpdateFunc return a TFJob object periodically.
+- Check `LastReconcileTime` to determine whether we should trigger a reconciler call.
+- `tf-operator` will list all pods/services which related to this TFJob.
+  + Compare the current state to the spec of this TFJob.
+  + Try to recovery the failed pod/service to make the training healthy.
+      + Error handing is described below.
+- Update the status of this TFJob.
+- TODO: we should call this reconciler with an exponential back-off delay (15s, 30s, 60s …) capped at 5 minutes.
+
+### Error Handling
+
+To make the system robust, the tf-operator should be able to locally and automatically recover from errors. 
+
+API v1alpha1 use an exit code semantics which is not very user-friendly:
+- User should keep the exit code (1-127, 128-255) in mind while they are coding.
+- tf-operator should look into the details of containers in each pod.
+
+**Option 1**
+
+Use kubernetes built-in `RestartPolicy` field to let user determine the restart policy:
+
+- Users set this field according to their model code.
+  + If set it to `OnFailure`, user should add reloading checkpoint code themselves.
+  + Otherwise restarting will take no effect.
+- It make more sense to let kubernetes to handing this restarting logic.
+- FYI, [ParameterServer restart crashes distributed training process](https://github.com/tensorflow/tensorflow/issues/7767#issuecomment-281841798).
+
+### Test
+
+**Unit Test**
+
+TBD
+
+**E2E Test**
+
+We can use this model from TensorFlow [repo](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tools/dist_test) for e2e test.
```

Comment:
Can you describe example e2e workflow? From beginning (empty jupyter) to the end (multinode tensorflow serving)?

I think it would be useful to have discussion (separate, dedicated doc?) how do we want to connect jupyter to tf-operator with user-experience in mind.

Another thing to consider which might affect this design is how do we transfer artifacts (code and trained models) between steps in workflow - I (data scientist) write code in jupyter -> somehow export it to place where tf-job (learning) can access it -> tf-job runs same code in distributed and scaled fashion -> tf-job saves model somewhere -> tf-serving runs in scaled fashion and consumes trained model.

Reply:
I think tf-operator does not know the jupyter since it is not in tf-operator's management. The operator only knows the YAML config file and I think the e2e test from jupyter to serving should be done in kubeflow/kubeflow.

Reply:
@inc0 Thanks for your high-level comments, it makes totally sense for me.

For e2e workflow we can re-use [kubernetes/test-infra](https://github.com/kubernetes/test-infra):

- Deploy a local/remote kubernetes cluster.
- Deploy TFJob controller and create CRD.
- Run e2e test case:
  - Load e2e test model from scratch.
  - Create TFJob yaml and deploy it.
  - Wait and check the result.

IMHO, put e2e design into a dedicated doc maybe better.
And i think TensorFlow serving is a very important feature of tf-operator, but i don't take something to think about it now :)